Add piecing it together practice exercise

zamora-carlos opened the pull request:
# pull request

<!-- Your content goes here: -->
Add _Piecing It Together_ practice exercise based on the instructions detailed [here](https://github.com/exercism/problem-specifications/tree/main/exercises/piecing-it-together).

In the `config.json` file, I included `exceptions`, `for-loops`, and `if-else-statements` as prerequisites. I wasn’t sure whether to add them to the `practices` array as well, since most exercises seem to leave it empty.

I’ve read the contribution guidelines, but please let me know if I’ve missed anything.

Closes #2949 


<!-- DO NOT EDIT BELOW THIS LINE! -->
---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/main/POLICIES.md#event-checklist)

zamora-carlos commented:
Hello, thank you for your detailed comments! I’ve made the requested changes. I just have one question: I currently placed the `JigsawInfo` class (previously `PartialJigsawInformation`) in both `.meta/src/reference/java` and `src/main/java`. Is that okay, or was it only needed in the reference folder?

kahgoh commented on exercises/practice/piecing-it-together/.meta/config.json:
```diff
@@ -0,0 +1,22 @@
+      ".meta/src/reference/java/PiecingItTogether.java"
+    ],
+    "editor": [
+      ".meta/src/reference/java/JigsawInfo.java"
```
This should point to the `src/main/java` one.

```suggestion
      "src/reference/java/JigsawInfo.java"
```

kahgoh reviewed (commented):
> I currently placed the `JigsawInfo` class (previously `PartialJigsawInformation`) in both `.meta/src/reference/java` and `src/main/java`. Is that okay, or was it only needed in the reference folder?

Yes, that's fine! In fact, that's currently needed for the CI tests to pass.

zamora-carlos commented on exercises/practice/piecing-it-together/src/main/java/JigsawInfo.java:
```diff
@@ -0,0 +1,131 @@
+    private final Optional<String> format;
+
+    private JigsawInfo(Builder builder) {
+        this.pieces = builder.pieces != null ? OptionalInt.of(builder.pieces) : OptionalInt.empty();
```
I knew that constructor didn’t look great, but I hadn’t thought about using Optionals and empty values in the builder class. Thanks for the suggestion. I’ve made the change, and the code looks much better now.

zamora-carlos commented:
Hi, I’ve updated the builder in the `JigsawInfo` class and fixed the style issues. The code now passes the verification check. Please let me know if any additional changes are needed.

kahgoh commented on exercises/practice/piecing-it-together/.meta/src/reference/java/PiecingItTogether.java:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        if (validGuesses.size() == 1) {
+            return validGuesses.getFirst();
```
CI gives the following error:

```
> Task :practice:piecing-it-together:compileJava FAILED
  (source = /home/runner/work/java/java/exercises/practice/piecing-it-together/.meta/src/reference/java/PiecingItTogether.java:/home/runner/work/java/java/exercises/practice/piecing-it-together/.meta/src/reference/java/JigsawInfo.java)
/home/runner/work/java/java/exercises/practice/piecing-it-together/.meta/src/reference/java/PiecingItTogether.java:44: error: cannot find symbol
            return validGuesses.getFirst();
                               ^
  symbol:   method getFirst()
  location: variable validGuesses of type List<JigsawInfo>
```

kahgoh reviewed (approved):
Looks good!
